### PR TITLE
Add SerializationTransformer

### LINKED ISSUE #48: Support serialization key transformer
In swift code we usually use camelCase while for JSON other convention (mostly snake_case) are used, I would don't like forcing the user to name the property we might have easy ways to support this:
## Detailed design

``` swift
SnakeCaseTransformer(User()) // convert the serialized object to snake case
```

This can be done by using `CustomSerializable`.
It would be good if the transformed can also be used to rename properties of an entity:

``` swift
struct User: Serializable {
   let id: Int
}

// you want id (Storable) to be named `userId` in the JSON

protocol KeyTransformer: CustomSerializable {
    func transformedKey(for candidate: String) -> String
}

extension KeyTransformer {
// default implementation that use `transformedKey...`to transform the keys
}
```

### PR DESCRIPTION
**Pull request by MP0w**
Closes #48 

@joanromano let me know what you think, I took a different approach than in feature/keyTransformer , while the other one didn't require touching the Implementation of Serializable/CustomSerializable and didn't require `CustomSerializable` to take care of the keyTransformer, this approach is much cleaner and easier to handle, other than performant

### CODE REVIEW

File: Source/JSONAPISerializer.swift [hunk 1]
```diff
@@ -67,33 +69,52 @@ public protocol JSONAPIEntity: CustomSerializable, JSONAPISerializable {
...
+
+/// A wrapper struct that handles the serialization of the JSON API include field
+private struct JSONAPIIncludedWrapper: CustomSerializable {
+    let object: JSONAPIConvertible
+    let includeChildren: Bool
+    
+    private func customSerialize(keyTransformer: KeyTransformer?) -> AnyObject? {
+        return object.includedRelationships(includeChildren, keyTransformer: keyTransformer)?.unifiedIncludedRelationships()
+    }
+}
```
Comment (joanromano) [inline Source/JSONAPISerializer.swift:91]: Interesting, what is the point of this wrappers?

Comment (MP0w) [inline Source/JSONAPISerializer.swift:91]: We need something like this because JSONAPIEntity is a strange `Serializable`. JSONAPIEntity's serialize method just call `object.data(includeRelationships ....` and `includedRelationships(...` is never in the serialized entity. 
`JSONAPIEntity` probably shouldn't even be a `Serializable` object (we might consider to remove the conformance).
Because of this I needed to handle the keyTransformer and I thought that the easiest and cleaner option would be to use `CustomSerializable` Wrappers that makes it really easy to handle the keyTransformer. 
Also now at initialization time nothing is serialized (while before we were creating the included field since it was not part of the serialization).

File: Tests/SerializationTransformerTests.swift [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                    let first = friends?.first as? [String: AnyObject]
+                    expect(first?.keys.first).to(equal("nAME"))
+                }
+                
+                do {
+                    let serialized = UppercaseTransformer(wrapped: LowercaseFirstCharacterTransformer(wrapped: friend)).serialize() as! [String: AnyObject]
+                    let friends = serialized["FRIENDS"] as? [AnyObject]
+                    let first = friends?.first as? [String: AnyObject]
+                    expect(first?.keys.first).to(equal("NAME"))
+                }
```
Comment (joanromano) [inline Tests/SerializationTransformerTests.swift:91]: Why this within a do?

Comment (MP0w) [inline Tests/SerializationTransformerTests.swift:91]: Just to reuse the code and variable names 🤓 And doesn't look bad. Don't like it?

File: Source/SerializationTransformer.swift [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    ```
+    struct User {
+       let userName: String
+    }
+ 
+    let result = SnakecaseTransformer(wrapped: User(userName: "Alex")) // -> {"user_name": "Alex"}
+ 
+    ```
+ */
+public struct SnakecaseTransformer<Wrapped: Serializable>: SerializationTransformer {
```
Comment (joanromano) [inline Source/SerializationTransformer.swift:66]: Do you really need to add the generic constraint here?

File: Tests/SerializationTransformerTests.swift [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    }
+    
+    override func spec() {
+        
+        let user = User(name: "Alex")
+        let friend = Friend(friends: [user])
+        
+        describe("Serialization Transformers") {
+            it("transforms the keys of a Serializable objects") {
+                let serialized = UppercaseTransformer(wrapped: friend).serialize() as! [String: AnyObject]
```
Comment (joanromano) [inline Tests/SerializationTransformerTests.swift:72]: So from now on if we want to use transformer we will have to wrap our objects into them and then call serialize on the transformers?

File: Source/SerializationTransformer.swift [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+import Foundation
+
+/// A closure that given a String as input output the transformed String, used to transform keys of the json
+public typealias KeyTransformer = (key: String) -> String
+
+/**
+ *  A protocol that special `Serializable` objects can adopt to transform other `Serializable` objects.
+ *  At the moment only key transforms are supported, see `SnakecaseTransformer` for a concrete implementations
+ */
```
Comment (joanromano) [inline Source/SerializationTransformer.swift:18]: `At the moment only key transforms are supported` I think we need a better way to constraint this. For now we only have this line in the documentation and the `promise` that all of our extensions in foundation classes and JSONAPI will use this just for keys, but won't necessarily be always the case.

File: Tests/SerializationTransformerTests.swift (hunk 4 above)
Comment (MP0w) [inline Tests/SerializationTransformerTests.swift:72]: Wrap them yes but in real life serialize() can't be called, you return a serializable wrapped into a transformer in the route handler

Comment (joanromano) [inline Tests/SerializationTransformerTests.swift:72]: Ah yes, then the router calls serialize(). I mean whatever the client is will call serialize() at some moment on the transformer.

File: Source/SerializationTransformer.swift (hunk 5 above)
Comment (MP0w) [inline Source/SerializationTransformer.swift:18]: What I mean is that you can't implement value transformer for example, but we will never move a key transformer into a everything-transformer

File: Source/SerializationTransformer.swift (hunk 3 above)
Comment (MP0w) [inline Source/SerializationTransformer.swift:66]: To be consistent with all other I guess? You have refer a non-generic that just has a wrapped: Serializable?
Why?

File: Tests/SerializationTransformerTests.swift (hunk 4 above)
Comment (MP0w) [inline Tests/SerializationTransformerTests.swift:72]: Yes exactly

Comment (MP0w) [inline Tests/SerializationTransformerTests.swift:72]: Do you have other API design in mind?

File: Source/SerializationTransformer.swift (hunk 3 above)
Comment (joanromano) [inline Source/SerializationTransformer.swift:66]: I think you can define the associated object as Serializable, no? I was just wondering

File: Source/SerializationTransformer.swift (hunk 5 above)
Comment (joanromano) [inline Source/SerializationTransformer.swift:18]: Yes but this is because the extensions use `KeyTransformer` only for keys, so it's technically not constrained right?

File: Source/JSONAPISerializer.swift (hunk 1 above)
Comment (joanromano) [inline Source/JSONAPISerializer.swift:91]: Yep, I like it

File: Tests/SerializationTransformerTests.swift (hunk 4 above)
Comment (joanromano) [inline Tests/SerializationTransformerTests.swift:72]: No no, nothing against the design. I wanted to understand it 100%

File: Tests/SerializationTransformerTests.swift (hunk 2 above)
Comment (joanromano) [inline Tests/SerializationTransformerTests.swift:91]: Didn't know you could use them here

**Comment by joanromano** [comment]
Looks really convenient (thanks for the tests, which help a lot also), my only concern is with the just documented constraint regarding only transforming keys.

File: Source/SerializationTransformer.swift (hunk 5 above)
Comment (MP0w) [inline Source/SerializationTransformer.swift:18]: Not sure what you mean, can you explain me better? I think I'm misunderstanding

Comment (joanromano) [inline Source/SerializationTransformer.swift:18]: So now we use this to only transform keys, right? For instance Serializer uses it on dictionary extension or on `serialize` method to do it. The thing is, no one prevent clients to use `keyTransformer` for other things than keys, no? So that sentence on the protocol might be a bit confusing probably.

**Comment by MP0w** [comment]
"documented constraint regarding only transforming keys" its the only one supported , what do you mean?

File: Source/SerializationTransformer.swift (hunk 5 above)
Comment (MP0w) [inline Source/SerializationTransformer.swift:18]: Oh, i get it. On CustomSerializable you have to implement it yourself, the document explain how you should implement it but if you misuse it and transform stuff other than the key than I guess you know what you are dong and is the expected behavior

File: Source/SerializationTransformer.swift (hunk 3 above)
Comment (MP0w) [inline Source/SerializationTransformer.swift:66]: Sure, but with he generic it gets the real type, we did the same for other objects, might not be useful but doesn't have any downside (I guess)

File: Tests/SerializationTransformerTests.swift (hunk 2 above)
Comment (MP0w) [inline Tests/SerializationTransformerTests.swift:91]: do {} is the equivalent of {} in objc

File: Source/SerializationTransformer.swift [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+     - returns: Should return the transformed key, for example an `UppercaseTransformer` would return the uppercase key.
+     */
+    func transform(key key: String) -> String
+}
+
+extension SerializationTransformer {
+    
+    /**
+     `SerializationTransformer`by default serialize it's object using the given keyTransformer (if any) and it's own key transformer.
+     Its own keyTransformer must always be used before the one provided by the callers, so UppercaseTransformer(LowercaseTransformer(object)) would result in uppercase keys.
```
Comment (joanromano) [inline Source/SerializationTransformer.swift:41]: This is a bit confusing: didn't you mean that his own `transform` must be used before the `KeyTransformer` provided by the callers?

Comment (joanromano) [inline Source/SerializationTransformer.swift:41]: And also the example `UppercaseTransformer(LowercaseTransformer(object))` does not provide any `KeyTransformer` right? It would be first a lowercase `transform` followed by an uppercase `transform` method.

Comment (joanromano) [inline Source/SerializationTransformer.swift:41]: Ok, I get it now:

`LowercaseTransformer` will have a `KeyTransformer` that is basically the `UppercaseTransformer` transform method, apart from his own `transform method` used first. 

Whereas `UppercaseTransformer`, in first term, won't have any `KeyTransformer` but will just return his transform method, the one that will be used afterwards by `LowercaseTransformer`.

Probably we could rephrase it to make it more understandable? Not quite sure how to do it though.

File: Tests/SerializationTransformerTests.swift [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            context("JSON API Links") {
+                it("should transform the keys") {
+                    let object = JSONAPILink.Object(href: "test", meta: friend)
+                    let serialized = UppercaseTransformer(wrapped: object).serialize() as! [String: AnyObject]
+                    expect(serialized["href"]).toNot(beNil())
+                    let meta = serialized["meta"] as? [String: AnyObject]
+                    expect(meta?["FRIENDS"]).toNot(beNil())
+                }
+            }
+        }
```
Comment (joanromano) [inline Tests/SerializationTransformerTests.swift:213]: Can we add a test that directly uses serialize of a transformer passing a `KeyTransformer`? I know that this is implicitly tested by wrapping a transformer into a transformer, but I think adding this one also would be nice.

File: Source/SerializationTransformer.swift (hunk 5 above)
Comment (joanromano) [inline Source/SerializationTransformer.swift:18]: Actually seeing again the sentence, I think this is not actually what this `SerializationTransformer` are being used for I think: right now the objects who use this protocol can only be used to wrap Serializable objects -> by calling serialize on this transformer objects, we will serialize the underlying serializable objects by passing the `KeyTransformer` closure, which will convert the keys.

File: Source/SerializationTransformer.swift (hunk 6 above)
Comment (MP0w) [inline Source/SerializationTransformer.swift:41]: Yep, had an hard time trying to explain it, I was hoping the example would clarify

File: Tests/SerializationTransformerTests.swift (hunk 7 above)
Comment (MP0w) [inline Tests/SerializationTransformerTests.swift:213]: We have transformer in transformer in: "should apply transformations in the right order".
IMO we shouldn't do it directly because doesn't really matter, that is a detail implementation of CustomSerializable for SerializationTransformers but people using kakapo shouldn't care about it because it's already done by kakapo with the protocol extension. Right?

File: Source/SerializationTransformer.swift (hunk 5 above)
Comment (joanromano) [inline Source/SerializationTransformer.swift:18]: Ok, just looked at it again. It's actually right, I would just modify it with `can adopt to transform other _wrapped_`Serializable`objects.`

**Comment by joanromano** [comment]
👍 awesome implementation, let's just work more on how to properly document now
